grails-app/services/org/pih/warehouse/product/ProductMergeService.groovy

```diff
@@ -0,0 +1,516 @@
+/**
+* Copyright (c) 2012 Partners In Health.  All rights reserved.
+* The use and distribution terms for this software are covered by the
+* Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
+* which can be found in the file epl-v10.html at the root of this distribution.
+* By using this software in any fashion, you are agreeing to be bound by
+* the terms of this license.
+* You must not remove this notice, or any other, from this software.
+**/ 
+package org.pih.warehouse.product
+
+import org.pih.warehouse.api.AvailableItem
+import org.pih.warehouse.core.Constants
+import org.pih.warehouse.inventory.Inventory
+import org.pih.warehouse.inventory.InventoryItem
+import org.pih.warehouse.inventory.Transaction
+import org.pih.warehouse.inventory.TransactionEntry
+import org.pih.warehouse.inventory.TransactionType
+import org.pih.warehouse.invoice.InvoiceItem
+import org.pih.warehouse.order.OrderItem
+import org.pih.warehouse.picklist.PicklistItem
+import org.pih.warehouse.receiving.ReceiptItem
+import org.pih.warehouse.requisition.RequisitionItem
+import org.pih.warehouse.shipping.ShipmentItem
+
+class ProductMergeService {
+
+    def inventoryService
+    def invoiceService
+    def productAvailabilityService
+    def requisitionService
+
+    /**
+     * Preforms product swapping for a product pairs passed as params
+     * */
+    def mergeProduct(String primaryId, String obsoleteId) {
```

Comment:
At some point, we talked about refactoring this into smaller methods. Since that's probably not going to happen before the merge, we should create a tech debt ticket to do that refactoring in the future.

If it's not too big of a task I would recommend we do it just to make this more readable and maintainable. 
```
validateProducts()
logProductMergeData()
mergeProductSuppliers()
mergeProductComponents()
mergeTransactions()
mergeRequisitions()
mergeShipments()
...
```

Even the validate method could be broken up into smaller methods.

Reply:
There's already a comment with the ticket number inside. I'll move it here, for visibility.